plutus-core/executables/Common.hs

```diff
@@ -559,3 +568,66 @@ runPrintExample getFn (ExampleOptions (ExampleSingle name)) = do
     T.putStrLn $ case lookup name examples of
         Nothing -> "Unknown name: " <> name
         Just ex -> PP.render $ prettyExample ex
+
+
+---------------- Print the cost model parameters ----------------
+
+runDumpModel :: IO ()
+runDumpModel = do
+    let params = fromJust PLC.defaultCostModelParams
+    BSL.putStr $ Aeson.encode params
+
+
+---------------- Print the type signatures of the default builtins ----------------
+
+-- Some types to represent signatures of built-in functions
+type PlcTerm = PLC.Term PLC.TyName PLC.Name PLC.DefaultUni PLC.DefaultFun ()
+type PlcType = PLC.Type PLC.TyName PLC.DefaultUni ()
+data QVarOrType = QVar String | Type PlcType  -- Quantified type variable or actual type
+
+data Signature = Signature [QVarOrType] PlcType  -- Argument types, return type
+instance Show Signature where
+    show (Signature args res) =
+        "[ " ++ (intercalate ", " $ map showQT args) ++ " ] -> " ++ showTy (normTy res)
+            where showQT =
+                      \case
+                       QVar tv -> "forall " ++ tv
+                       Type ty -> showTy (normTy ty)
+                  normTy :: PlcType -> PlcType
+                  normTy ty = PLC.runQuote $ PLC.unNormalized <$> normalizeType ty
+                  showTy ty =
+                      case ty of
+                        PLC.TyBuiltin _ t -> show $ PP.pretty t
+                        PLC.TyApp {}      -> showMultiTyApp $ unwrapTyApp ty
+                        _                 -> show $ PP.pretty ty
+                  unwrapTyApp ty =
+                      case ty of
+                        PLC.TyApp _ t1 t2 -> (unwrapTyApp t1) ++ [t2]
+                        -- Assumes iterated built-in type applications all associate to the left;
+                        -- if not, we'll just get some odd formatting.
+                        _                 -> [ty]
```

Comment:
Have you considered using the `Readable` pretty-printer? It really is supposed to give you a readable output. Can't run it in my head, but it could fix all the problems you're encountering and then some (for example, render stuff as `list a` rather than `list (a)`).